### DISCUSSION TITLE
FIELD_TYPE.JSON ignored

### DISCUSSION THREAD

**Discussion by jojozg7** [PyMySQL/mysqlclient#646] [Issue report]
### Describe the bug

Columns of type json are ignored.

### Environment

Debian Bookworm, Python 3.11

### How did you install libmysqlclient libraries?

python3-mysqldb

### What version of mysqlclient do you use?

1.4.6

### Minimum but complete code to reproduce

```python
# Write Python code here.
import MySQLdb
from MySQLdb.constants import FIELD_TYPE
import json

def to_json(value):
    return json.loads(value)

conn = MySQLdb.connect(
    user="root",
    port=3306.
    host="127.0.0.1",
    conv={ FIELD_TYPE.JSON: to_json }
)

cur = conn.cursor()
cur.execute("select * from test")
print(cur.fetchone())
```

### Schema and initial data required to reproduce.

```sql
-- Write SQL here.
-- e.g. CREATE TABLE ...
create table test (test json);
insert into test (test) values ('{"key":"value"}');
```

**Comment by methane** [comment]
Its intentional.
Please don't try to use converter from applications.
Driver is low level library. Use ORM for automatic conversion.

**Comment by jojozg7** [comment]
Why is it documented here in the USER GUIDE if you are not supposed to use it?

https://github.com/PyMySQL/mysqlclient/blob/main/doc/user_guide.rst

**Reply by methane** [reply]
They are used by ORMs, like Django customize how time columns are converted.
JSON and other str/bytes columns are not supported.

Why it is not documented is I am lazy, volunteer maintainer.
This project was dead once. I maintain to keep working. But I don't maintain document, because I am not good at English.

**Comment by jojozg7** [comment]
mysql-connector (https://www.mysql.com/products/connector/), mariadb and psycopg2 offer the possibility in exactly the same way. I am not quite sure why this should not be possible here.

**Reply by methane** [reply]
I don't think they are "exactly same way". Do all of them accept `conv` dict and functions in the dict accept str? How they treat encoidng or other additional information needed for conversion?

This library has very long history. The API designed in Python 2 era.
I port this library to Python 3 and consume very hard time for fixing Unicode errors with minimal backward incompatibility.

`conv` parameter is the most hard API. I wanted to drop it completely, but I desided to keep it for customizing how decimal, datetime are converted. Since they are converted in driver, ORM or user can not restore the information lost during the conversion in the driver. Additionally, Decimal and datetime are represented in ASCII. So no encoding is needed.

Instead, I decided to drop support only str/bytes types in `conv`. They are conplex because they need to know field_type, flags, and encodings. Unlike decimal or datetime, user or higher level library can convert them later without data loss.

Lastly, I am tired to maintain this library. When I started to maintain this, there are no Python 3 MySQL driver. Now they are mysql-connector-python and mariadb.
I don't want to increase maintain cost. I want to decrease. This is the most important reason why it is not possible.

If you don't understand this, you can fork this library and add that feature and maintain perfect document.
